### PR TITLE
dptp-cm: Use flagutil.KubernetesOptions

### PR DESCRIPTION
**Pull request by hongkailiu**
Follow up https://github.com/openshift/ci-tools/pull/2311#pullrequestreview-745345438

/cc @stevekuznetsov

### CODE REVIEW

File: cmd/dptp-controller-manager/main.go
Comment (stevekuznetsov) [inline cmd/dptp-controller-manager/main.go:246]: Do we need this branch? What do we do in the deployment? I don't understand why our loading has to be so complex. Either always expect it to be in-cluster, or always expect it to be present in the configs ...

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:246]: It turns out we need both:

development: load from $KUBECONFIG
https://github.com/openshift/ci-tools/blob/12fad0a042c90eda3cddf44fb8dd84a1ea6ea9e6/hack/run-dptp-controller-manager.sh#L11

production: load from inClusterConfig because the secret does not include the one for `app.ci`:

https://github.com/openshift/release/blob/fa66cd849fb407495cd50fab336d3f0e88fcb622/core-services/ci-secret-bootstrap/_config.yaml#L1948-L1961

Maybe add it into the secret and then remove the code for inClusterConfig?

Comment (stevekuznetsov) [reply cmd/dptp-controller-manager/main.go:246]: Change development. We do not write code to handle development situations, we write it to handle production. Developing on different code-paths to production is a recipe for disaster.

Comment (stevekuznetsov) [reply cmd/dptp-controller-manager/main.go:246]: The k8s loader also allows you to provide it a loading strategy, so that it perfers one to the other. @alvaroaleman is there any simpler way to do this? I really dislike how damn complicated we've made loading. The library is complex as all get-out, then each caller does stuff like this to make it even more impossible.

Comment (alvaroaleman) [reply cmd/dptp-controller-manager/main.go:246]: This uses the in-cluster config for app.ci if there is no app.ci config. I don't really see how anything in client-go would be able to do this, as it doesn't know what name the in-cluster config is for.

I also don't really see how any change in the deployment would improve things here. Sure, we could add a flag for `--default-in-cluster-config-for-cluster-name` but I don't anticipate this changing so it looks more like busywork than like something that improves things to me.

Comment (stevekuznetsov) [reply cmd/dptp-controller-manager/main.go:246]: I just hate that we have 35 places that load config, and their call-sites are all over the place. Half of them do something like this, half don't, half have custom handling for different things, half don't ... next time anyone has to do a broad change it's going to be as much of a nightmare as it has been for Hongkai. Surely we can do better?

Comment (stevekuznetsov) [reply cmd/dptp-controller-manager/main.go:246]: Maybe load in-cluster if possible, add it to a well-known key in the output mapping that the upstream library provides, then the caller just moves map keys around?

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:252]: Updated.
Not sure if it is getting better or worse though.

Comment (stevekuznetsov) [reply cmd/dptp-controller-manager/main.go:252]: Do we put the in-cluster config into a known key? Can we fail over to that? @alvaroaleman any thoughts?

Comment (alvaroaleman) [reply cmd/dptp-controller-manager/main.go:252]: Yeah we do put it into a known key, it is handled above

File: cmd/dptp-controller-manager/main.go
Comment (alvaroaleman) [inline cmd/dptp-controller-manager/main.go:257]: That is never going to pass, the Host will be `$KUBERNETES_SERVICE_HOST + KUBERNETES_SERVICE_PORT" which is an ip address: port (and probably the same among all openshift clusters. I would just remove the `else` clause

Comment (alvaroaleman) [reply cmd/dptp-controller-manager/main.go:257]: /hold

File: cmd/dptp-controller-manager/main.go
Comment (alvaroaleman) [inline cmd/dptp-controller-manager/main.go:303]: You can not pass a pointer to the loop var

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:307]: Sorry. One more thing that I need to remember.

File: cmd/dptp-controller-manager/main.go
Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go]: Updated. Removing the `else` may break the dev script.

File: cmd/dptp-controller-manager/main.go
Comment (alvaroaleman) [inline cmd/dptp-controller-manager/main.go:259]: But why? This code shouldn't reimplement part of the client-go logic (and break if client-go should be changed). So far, we trusted that a context named `app.ci` is for app.ci and that the in-cluster config is for that as well. What is the reason not to trust that anymore?

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:259]: For example, if we accidentally removed this line

https://github.com/openshift/ci-tools/blob/cc37aa0f41d5e5d71739d87123433e6d34ec7d9c/hack/run-dptp-controller-manager.sh#L11

then we would use another cluster as `app.ci`.

I just want to be safer.

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:259]: Previously, we load `inCluster` without the logic from upstream and now the upstream logic could give me a wrong result.

Comment (alvaroaleman) [reply cmd/dptp-controller-manager/main.go:259]: Why would in the example give a wrong result?

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go:259]: It could come from `currentContext` of the last kubeconfig file it loads.

https://github.com/kubernetes/test-infra/blob/c3f3e307169f26b3790a132060c4bb2f3c76a66a/prow/kube/config.go#L58

and

https://github.com/kubernetes/test-infra/blob/c3f3e307169f26b3790a132060c4bb2f3c76a66a/prow/kube/config.go#L121-L135

Comment (alvaroaleman) [reply cmd/dptp-controller-manager/main.go:259]: Ok, I am just uneasy about relying on implementation details here. The thing needs to be prefixed with `https://` in the in-cluster case and joined with net.JoinHostPorT: https://github.com/openshift/release-controller/blob/efd4b6d1c92ff73804e397524de974d160a8dc31/vendor/k8s.io/client-go/rest/config.go#L508

Comment (hongkailiu) [reply cmd/dptp-controller-manager/main.go]: Updated.

**Review by alvaroaleman — APPROVED** [review]

**Comment by hongkailiu** [comment]
/hold cancel

I will be watching ...

**Comment by hongkailiu** [comment]
```
oc get pod -n ci dptp-controller-manager-9fc689774-th5hw
NAME                                      READY   STATUS    RESTARTS   AGE
dptp-controller-manager-9fc689774-th5hw   1/1     Running   0          20m
```
seems working

**Pull request merged**
